### PR TITLE
qt: Fix default font weight values if selected values aren't supported by the font

### LINKED ISSUE #3848: Wallet won't stay open Linux
I upgraded wallet from Linux 15 to dashcore-0.16.1.1-x86_64-linux-gnu.tar.gz wallet.  Running ./dash-qt attempts to start and immediately closes.

To reproduce:
1. Open wallet and try again, not working.
2. Backup and delete ~/.dashcore and ~/.config/Dash folders
3. Retry, same thing
The create or restore wallet windows opens, I select new wallet, then the splash screen is visible, after a short while the wallet closes (and not running in the background)

The terminal output when this happens is:

```[user@pc bin]$ ./dash-qt
Assertion failure:
  assertion: false
  file: qt/guiutil.cpp, line: 1771
  function: int GUIUtil::supportedWeightToIndex(QFont::Weight)
   0#: (0x55ACC53C4BA6) <unknown-file> - ???
   1#: (0x55ACC4C4775E) <unknown-file> - ???
   2#: (0x55ACC4C6D025) <unknown-file> - ???
   3#: (0x55ACC58AAD30) <unknown-file> - ???
   4#: (0x55ACC58CB8B3) <unknown-file> - ???
   5#: (0x55ACC58CC2FF) <unknown-file> - ???
   6#: (0x55ACC58CC40D) <unknown-file> - ???
   7#: (0x55ACC4C4C703) <unknown-file> - ???
   8#: (0x55ACC4C09C3C) <unknown-file> - ???
   9#: (0x55ACC4C0A2AC) <unknown-file> - ???
  10#: (0x55ACC5FE5382) <unknown-file> - ???
  11#: (0x55ACC59F1CD4) <unknown-file> - ???
  12#: (0x55ACC56956EF) <unknown-file> - ???
  13#: (0x55ACC568FE3C) <unknown-file> - ???
dash-qt: qt/guiutil.cpp:1771: int GUIUtil::supportedWeightToIndex(QFont::Weight): Assertion `false' failed.
Aborted (core dumped)
[user@pc bin]$ ```

I have also tried the following commands after deleting the same folders above (clean install) which were recommended via help from Discord:  

`dash-qt -resetguisettings`
`dash-qt -font-weight-normal=0 -font-weight-bold=8`
`dash-qt -font-weight-normal=0 -font-weight-bold=0`

All of these produce the same result

### Machine specs:

- OS: Linux (Manjaro/Arch) with XFCE
- CPU: Intel(R) Core(TM) i7-2600K CPU @ 3.40GHz
- RAM: 32g
- Disk size: 10tb
- Disk Type (HD/SDD): hdd

### Extra Information

Debug log show same message.

### PR DESCRIPTION
**Pull request by xdustinface**
Fixes #3848

### CODE REVIEW

**Review by UdjinM6 — CHANGES_REQUESTED** [review]
So basically it's smth like "should guess font weights for all font families, not only the default one, and also do this regardless of the number of font weights they support". Right?

Also, pls see one tiny issue below.

File: src/qt/guiutil.cpp [hunk 1]
```diff
@@ -302,13 +302,25 @@ void setupAmountWidget(QLineEdit *widget, QWidget *parent)
...
-        if (fontFamily == FontFamily::SystemDefault && getSupportedWeights().size() < 4) {
-            fontWeightNormal = mapSupportedWeights[FontFamily::SystemDefault].front();
-            fontWeightBold = mapSupportedWeights[FontFamily::SystemDefault].back();
-            QSettings().setValue("fontWeightNormal", weightToArg(fontWeightNormal));
-            QSettings().setValue("fontWeightBold", weightToArg(fontWeightBold));
+        std::vector<QFont::Weight> vecSupportedWeights = getSupportedWeights();
+        // See if the default value for normal weight is available
+        auto it = std::find(vecSupportedWeights.begin(), vecSupportedWeights.end(), defaultFontWeightNormal);
+        if (it == vecSupportedWeights.end()) {
+            // If not, use the second lightest available weight as bold weight default or the lightest if there is only one
```
[the remark this answers is not in the document]
Comment (xdustinface) [reply src/qt/guiutil.cpp]: Fixed in  eb42b36

**Review by UdjinM6 — APPROVED** [review]
Slightly tested ACK

(doesn't break anything on my mac)

**Comment by xdustinface** [comment]
> So basically it's smth like "should guess font weights for all font families, not only the default one, and also do this regardless of the number of font weights they support". Right?
> 

Ops, did not see this comment before 🙈 

Yeah now it's done for all fonts selected after running first time or running with `-resetguisetting` which probably wouldn't be required but doesn't hurt in the end. The problem in the issue was obviously that the OS default font didn't support one or both of the font weights we have as default but still had more than 4 weights so the adjustments i reworked there didn't correct the weight values to the available ones.. hard to say for sure but i think it should be that case 🤔

**Comment by PastaPastaPasta** [comment]
Is this a potential backport candidate?

File: src/qt/guiutil.cpp [hunk 2]
```diff
@@ -302,13 +302,25 @@ void setupAmountWidget(QLineEdit *widget, QWidget *parent)
...
-        // First make sure SystemDefault has reasonable default values if it does not support the full range of weights.
-        if (fontFamily == FontFamily::SystemDefault && getSupportedWeights().size() < 4) {
-            fontWeightNormal = mapSupportedWeights[FontFamily::SystemDefault].front();
-            fontWeightBold = mapSupportedWeights[FontFamily::SystemDefault].back();
-            QSettings().setValue("fontWeightNormal", weightToArg(fontWeightNormal));
-            QSettings().setValue("fontWeightBold", weightToArg(fontWeightBold));
+        std::vector<QFont::Weight> vecSupportedWeights = getSupportedWeights();
+        // See if the default value for normal weight is available
+        auto it = std::find(vecSupportedWeights.begin(), vecSupportedWeights.end(), defaultFontWeightNormal);
+        if (it == vecSupportedWeights.end()) {
```
Comment (PastaPastaPasta) [inline src/qt/guiutil.cpp:308]: Should probably combine these no? `it` isn't actually used for anything

File: src/qt/guiutil.cpp [hunk 3]
```diff
@@ -302,13 +302,25 @@ void setupAmountWidget(QLineEdit *widget, QWidget *parent)
...
+            // If not, use the lightest available weight as normal weight
+            fontWeightNormal = vecSupportedWeights.front();
         }
+
+        // See if the default value for bold weight is available
+        it = std::find(vecSupportedWeights.begin(), vecSupportedWeights.end(), defaultFontWeightBold);
+        if (it != vecSupportedWeights.end()) {
```
Comment (PastaPastaPasta) [inline src/qt/guiutil.cpp:315]: Same, combine

**Review by PastaPastaPasta — CHANGES_REQUESTED** [review]

File: src/qt/guiutil.cpp (hunk 3 above)
Comment (UdjinM6) [inline src/qt/guiutil.cpp:315]: Wait... this should actually be
```suggestion
        if (it == vecSupportedWeights.end()) {
```
I think

Comment (xdustinface) [reply src/qt/guiutil.cpp:315]: Yeah also realized it :D

**Review by UdjinM6 — APPROVED** [review]
utACK
(still works on my mac :) )

**Review by PastaPastaPasta — APPROVED** [review]
utACK

**Pull request merged**